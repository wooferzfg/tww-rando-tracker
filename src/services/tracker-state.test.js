import _ from 'lodash';

import Locations from './locations';
import LogicHelper from './logic-helper';
import TrackerState from './tracker-state';

describe('TrackerState', () => {
  beforeEach(() => {
    Locations.reset();
    LogicHelper.reset();
  });

  describe('default', () => {
    beforeEach(() => {
      const itemLocations = {
        'Outset Island': {
          'Savage Labyrinth - Floor 30': {
            test: 'data',
          },
        },
        'Dragon Roost Cavern': {
          "Bird's Nest": {
            test: 'data',
          },
        },
      };
      const locationsList = [
        {
          generalLocation: 'Outset Island',
          detailedLocation: 'Savage Labyrinth - Floor 30',
        },
        {
          generalLocation: 'Dragon Roost Cavern',
          detailedLocation: "Bird's Nest",
        },
      ];
      Locations.initializeRaw(itemLocations, locationsList);

      LogicHelper.startingItems = {
        'Progressive Shield': 1,
        'Wind Waker': 1,
        "Boat's Sail": 1,
        "Wind's Requiem": 1,
        'Ballad of Gales': 1,
        'Song of Passing': 1,
        'Progressive Sword': 1,
      };
    });

    test('initializes the entrances as an empty object', () => {
      const defaultState = TrackerState.default();

      expect(defaultState.entrances).toEqual({});
    });

    test('initializes the items', () => {
      const defaultState = TrackerState.default();

      expect(defaultState.items).toMatchSnapshot();
    });

    test('initializes the checked locations', () => {
      const defaultState = TrackerState.default();

      expect(defaultState.locationsChecked).toMatchSnapshot();
    });

    test('initializes the items for locations', () => {
      const defaultState = TrackerState.default();

      expect(defaultState.itemsForLocations).toMatchSnapshot();
    });
  });

  describe('createStateRaw', () => {
    test('creates the state with the provided entrances, items, and locations checked', () => {
      const items = {
        'Deku Leaf': 2,
      };
      const entrances = {
        'Needle Rock Isle Secret Cave': 'Dragon Roost Cavern',
      };
      const locationsChecked = {
        'Dragon Roost Cavern': {
          "Bird's Nest": true,
        },
      };
      const itemsForLocations = {
        'Dragon Roost Cavern': {
          "Bird's Nest": 'Deku Leaf',
        },
      };

      const newState = TrackerState.createStateRaw({
        entrances,
        items,
        itemsForLocations,
        locationsChecked,
      });

      expect(newState.entrances).toEqual(entrances);
      expect(newState.items).toEqual(items);
      expect(newState.itemsForLocations).toEqual(itemsForLocations);
      expect(newState.locationsChecked).toEqual(locationsChecked);
    });
  });

  describe('readState', () => {
    let expectedItems;
    let expectedEntrances;
    let expectedLocationsChecked;
    let expectedItemsForLocations;
    let trackerState;

    beforeEach(() => {
      expectedItems = {
        'Deku Leaf': 2,
      };
      expectedEntrances = {
        'Needle Rock Isle Secret Cave': 'Dragon Roost Cavern',
      };
      expectedLocationsChecked = {
        'Dragon Roost Cavern': {
          "Bird's Nest": true,
        },
      };
      expectedItemsForLocations = {
        'Dragon Roost Cavern': {
          "Bird's Nest": 'Deku Leaf',
        },
      };

      trackerState = TrackerState.createStateRaw({
        entrances: expectedEntrances,
        items: expectedItems,
        itemsForLocations: expectedItemsForLocations,
        locationsChecked: expectedLocationsChecked,
      });
    });

    test('returns an object with the entire state', () => {
      const stateData = trackerState.readState();

      expect(stateData).toEqual({
        entrances: expectedEntrances,
        items: expectedItems,
        itemsForLocations: expectedItemsForLocations,
        locationsChecked: expectedLocationsChecked,
      });
    });
  });

  describe('getItemValue', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();
      state.items = {
        'Deku Leaf': 2,
      };
    });

    test('returns the value of the item', () => {
      const itemValue = state.getItemValue('Deku Leaf');

      expect(itemValue).toEqual(2);
    });
  });

  describe('incrementItem', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();
    });

    describe('when the item is already at max quantity', () => {
      beforeEach(() => {
        LogicHelper.startingItems = {
          'Progressive Sword': 2,
        };

        state.items = {
          'Progressive Sword': 4,
        };
      });

      test('returns a new state with the item reset to the minimum quantity', () => {
        const newState = state.incrementItem('Progressive Sword');

        expect(newState.items['Progressive Sword']).toEqual(2);
      });
    });

    describe('when the item is not already at max quantity', () => {
      beforeEach(() => {
        state.items = {
          'Deku Leaf': 0,
        };
      });

      test('returns a new state with the item count incremented by 1', () => {
        const newState = state.incrementItem('Deku Leaf');

        expect(newState.items['Deku Leaf']).toEqual(1);
      });
    });
  });

  describe('decrementItem', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();
    });

    describe('when the item is already at min quantity', () => {
      beforeEach(() => {
        LogicHelper.startingItems = {
          'Progressive Sword': 2,
        };

        state.items = {
          'Progressive Sword': 2,
        };
      });

      test('returns a new state with the item reset to the maximum quantity', () => {
        const newState = state.decrementItem('Progressive Sword');

        expect(newState.items['Progressive Sword']).toEqual(4);
      });
    });

    describe('when the item is not already at min quantity', () => {
      beforeEach(() => {
        state.items = {
          'Deku Leaf': 1,
        };
      });

      test('returns a new state with the item count decremented by 1', () => {
        const newState = state.decrementItem('Deku Leaf');

        expect(newState.items['Deku Leaf']).toEqual(0);
      });
    });
  });

  describe('getEntranceForExit', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();
      state.entrances = {
        'Dragon Roost Cavern': 'Needle Rock Isle Secret Cave',
      };
    });

    test('returns the value of the entrance', () => {
      const entranceValue = state.getEntranceForExit('Needle Rock Isle Secret Cave');

      expect(entranceValue).toEqual('Dragon Roost Cavern');
    });
  });

  describe('getExitForEntrance', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();
      state.entrances = {
        'Dragon Roost Cavern': 'Needle Rock Isle Secret Cave',
      };
    });

    test('returns the value of the entrance', () => {
      const entranceValue = state.getExitForEntrance('Dragon Roost Cavern');

      expect(entranceValue).toEqual('Needle Rock Isle Secret Cave');
    });
  });

  describe('setExitForEntrance', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();
      state.entrances = {
        'Dragon Roost Cavern': 'Needle Rock Isle Secret Cave',
        'Ice Ring Isle Secret Cave': 'Forbidden Woods',
      };
      state.items = {
        'Entered Needle Rock Isle Cave': 1,
        'Entered FW': 1,
        'Entered Ice Ring Isle Cave': 0,
      };
    });

    test('returns a new state with the entrance value and entry item modified', () => {
      const newState = state.setExitForEntrance('Forbidden Woods', 'Ice Ring Isle Secret Cave');

      expect(newState.entrances).toEqual({
        'Dragon Roost Cavern': 'Needle Rock Isle Secret Cave',
        'Ice Ring Isle Secret Cave': 'Forbidden Woods',
        'Forbidden Woods': 'Ice Ring Isle Secret Cave',
      });
      expect(newState.items).toEqual({
        'Entered Needle Rock Isle Cave': 1,
        'Entered FW': 1,
        'Entered Ice Ring Isle Cave': 1,
      });
    });

    test('when marking an entrance that leads to nothing, does not modify items', () => {
      const newState = state.setExitForEntrance('Forbidden Woods', LogicHelper.NOTHING_EXIT);

      expect(newState.entrances).toEqual({
        'Dragon Roost Cavern': 'Needle Rock Isle Secret Cave',
        'Ice Ring Isle Secret Cave': 'Forbidden Woods',
        'Forbidden Woods': LogicHelper.NOTHING_EXIT,
      });
      expect(newState.items).toEqual({
        'Entered Needle Rock Isle Cave': 1,
        'Entered FW': 1,
        'Entered Ice Ring Isle Cave': 0,
      });
    });
  });

  describe('unsetExit', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();
      state.entrances = {
        'Dragon Roost Cavern': 'Needle Rock Isle Secret Cave',
        'Tower of the Gods': 'Forbidden Woods',
      };
      state.items = {
        'Entered Needle Rock Isle Cave': 1,
        'Entered FW': 1,
      };
    });

    test('returns a new state with the entrance value and entry item unset', () => {
      const newState = state.unsetExit('Needle Rock Isle Secret Cave');

      expect(newState.entrances).toEqual({
        'Tower of the Gods': 'Forbidden Woods',
      });
      expect(newState.items).toEqual({
        'Entered Needle Rock Isle Cave': 0,
        'Entered FW': 1,
      });
    });
  });

  describe('unsetEntrance', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();
      state.entrances = {
        'Dragon Roost Cavern': 'Needle Rock Isle Secret Cave',
        'Tower of the Gods': 'Forbidden Woods',
        'Cliff Plateau Isles Secret Cave': LogicHelper.NOTHING_EXIT,
      };
      state.items = {
        'Entered Needle Rock Isle Cave': 1,
        'Entered FW': 1,
      };
    });

    test('returns a new state with the entrance value and entry item unset', () => {
      const newState = state.unsetEntrance('Dragon Roost Cavern');

      expect(newState.entrances).toEqual({
        'Tower of the Gods': 'Forbidden Woods',
        'Cliff Plateau Isles Secret Cave': LogicHelper.NOTHING_EXIT,
      });
      expect(newState.items).toEqual({
        'Entered Needle Rock Isle Cave': 0,
        'Entered FW': 1,
      });
    });

    test('when unsetting an entrance that leads to nothing, does not modify items', () => {
      const newState = state.unsetEntrance('Cliff Plateau Isles Secret Cave');

      expect(newState.entrances).toEqual({
        'Dragon Roost Cavern': 'Needle Rock Isle Secret Cave',
        'Tower of the Gods': 'Forbidden Woods',
      });
      expect(newState.items).toEqual({
        'Entered Needle Rock Isle Cave': 1,
        'Entered FW': 1,
      });
    });
  });

  describe('isEntranceChecked', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();
      state.entrances = {
        'Dragon Roost Cavern': 'Needle Rock Isle Secret Cave',
        'Tower of the Gods': 'Forbidden Woods',
      };
    });

    describe('when the entrance is checked', () => {
      test('returns true', () => {
        const isEntranceChecked = state.isEntranceChecked('Dragon Roost Cavern');

        expect(isEntranceChecked).toEqual(true);
      });
    });

    describe('when the entrance is not checked', () => {
      test('returns false', () => {
        const isEntranceChecked = state.isEntranceChecked('Forbidden Woods');

        expect(isEntranceChecked).toEqual(false);
      });
    });
  });

  describe('isLocationChecked', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();
      state.locationsChecked = {
        'Dragon Roost Cavern': {
          "Bird's Nest": true,
        },
      };
    });

    test('returns whether the location is checked', () => {
      const isLocationChecked = state.isLocationChecked('Dragon Roost Cavern', "Bird's Nest");

      expect(isLocationChecked).toEqual(true);
    });
  });

  describe('toggleLocationChecked', () => {
    let state;

    beforeEach(() => {
      const initialLocationsChecked = {
        'Dragon Roost Cavern': {
          "Bird's Nest": true,
        },
      };

      state = new TrackerState();
      state.locationsChecked = _.cloneDeep(initialLocationsChecked);
    });

    describe('when the location is already checked', () => {
      beforeEach(() => {
        state.locationsChecked = {
          'Dragon Roost Cavern': {
            "Bird's Nest": true,
          },
        };
      });

      test('returns a new state with the location not checked', () => {
        const newState = state.toggleLocationChecked('Dragon Roost Cavern', "Bird's Nest");

        const newIsLocationChecked = _.get(newState.locationsChecked, ['Dragon Roost Cavern', "Bird's Nest"]);

        expect(newIsLocationChecked).toEqual(false);
      });
    });

    describe('when the location is not already checked', () => {
      beforeEach(() => {
        state.locationsChecked = {
          'Dragon Roost Cavern': {
            "Bird's Nest": false,
          },
        };
      });

      test('returns a new state with the location not checked', () => {
        const newState = state.toggleLocationChecked('Dragon Roost Cavern', "Bird's Nest");

        const newIsLocationChecked = _.get(newState.locationsChecked, ['Dragon Roost Cavern', "Bird's Nest"]);

        expect(newIsLocationChecked).toEqual(true);
      });
    });
  });

  describe('getLocationsForItem', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();
      state.itemsForLocations = {
        'Windfall Island': {
          'Maggie - Free Item': 'Progressive Sword',
          'House of Wealth Chest': 'Progressive Sword',
          'Tott - Play Rhythm': 'Bombs',
        },
        'Dragon Roost Cavern': {
          'First Room': 'Boomerang',
          "Bird's Nest": 'Progressive Sword',
        },
      };
    });

    test('returns the location for the item', () => {
      const location = state.getLocationsForItem('Progressive Sword');

      expect(location).toEqual([{
        generalLocation: 'Windfall Island',
        detailedLocation: 'Maggie - Free Item',
      }, {
        generalLocation: 'Windfall Island',
        detailedLocation: 'House of Wealth Chest',
      }, {
        generalLocation: 'Dragon Roost Cavern',
        detailedLocation: "Bird's Nest",
      }]);
    });
  });

  describe('getItemForLocation', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();
      state.itemsForLocations = {
        'Windfall Island': {
          'Maggie - Free Item': 'Deku Leaf',
        },
      };
    });

    test('returns the value of the entrance', () => {
      const entranceValue = state.getItemForLocation('Windfall Island', 'Maggie - Free Item');

      expect(entranceValue).toEqual('Deku Leaf');
    });
  });

  describe('setItemForLocation', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();

      state.itemsForLocations = {
        'Dragon Roost Cavern': {
          "Bird's Nest": null,
        },
      };
    });

    test('returns a new state with the item for location value modified', () => {
      const newState = state.setItemForLocation('Grappling Hook', 'Windfall Island', 'Maggie - Free Item');

      const newItemForLocation = _.get(newState.itemsForLocations, ['Windfall Island', 'Maggie - Free Item']);

      expect(newItemForLocation).toEqual('Grappling Hook');
    });
  });

  describe('getChartFromChartMapping', () => {
    test('returns chart', () => {
      const state = new TrackerState();

      state.islandsForCharts = {
        'Treasure Chart 24': 'Windfall Island',
      };

      const chart = state.getChartFromChartMapping('Windfall Island');

      expect(chart).toBe('Treasure Chart 24');
    });
  });

  describe('getIslandFromChartMapping', () => {
    test('returns island', () => {
      const state = new TrackerState();

      state.islandsForCharts = {
        'Treasure Chart 24': 'Windfall Island',
      };

      const island = state.getIslandFromChartMapping('Treasure Chart 24');

      expect(island).toBe('Windfall Island');
    });
  });

  describe('setChartMapping', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();

      state.islandsForCharts = {};
    });

    test('sets chart mapping', () => {
      const newState = state.setChartMapping('Treasure Chart 10', 'Chart for Outset Island');

      const newIslandsForCharts = _.get(newState.islandsForCharts, 'Treasure Chart 10');

      expect(newIslandsForCharts).toBe('Outset Island');
    });
  });

  describe('unsetChartMapping', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();

      state.islandsForCharts = {
        'Treasure Chart 23': 'Pawprint Isle',
      };
    });

    test('removes chart mapping', () => {
      const newState = state.unsetChartMapping('Chart for Pawprint Isle');

      const newIslandsForCharts = _.get(newState.islandsForCharts, 'Treasure Chart 23', null);

      expect(newIslandsForCharts).toBe(null);
    });

    test('does not remove chart mapping for different chart', () => {
      const newState = state.unsetChartMapping('Chart for Windfall Island');

      const newIslandsForCharts = _.get(newState.islandsForCharts, 'Treasure Chart 23', null);

      expect(newIslandsForCharts).toBe('Pawprint Isle');
    });
  });

  describe('unsetItemForLocation', () => {
    let state;

    beforeEach(() => {
      state = new TrackerState();

      state.itemsForLocations = {
        'Dragon Roost Cavern': {
          "Bird's Nest": 'Bombs',
        },
      };
    });

    test('returns a new state with the item for location value set to null', () => {
      const newState = state.unsetItemForLocation('Dragon Roost Cavern', "Bird's Nest");

      const newItemForLocation = _.get(newState.itemsForLocations, ['Dragon Roost Cavern', "Bird's Nest"]);

      expect(newItemForLocation).toEqual(null);
    });
  });

  describe('clearBannedLocations', () => {
    let state;

    beforeEach(() => {
      const itemLocations = {
        'Outset Island': {
          'Savage Labyrinth - Floor 30': {
            test: 'data',
          },
          'Savage Labyrinth - Floor 50': {
            test: 'data',
          },
          'Sunken Treasure': {
            test: 'data',
          },
        },
        'Dragon Roost Cavern': {
          'First Room': {
            test: 'data',
          },
          'Alcove With Water Jugs': {
            test: 'data',
          },
          "Bird's Nest": {
            test: 'data',
          },
        },
        'Forsaken Fortress': {
          'Phantom Ganon': {
            test: 'data',
          },
        },
        Mailbox: {
          'Letter from Aryll': {
            test: 'data',
          },
          'Letter from Tingle': {
            test: 'data',
          },
        },
      };
      const locationsList = [
        {
          generalLocation: 'Outset Island',
          detailedLocation: 'Savage Labyrinth - Floor 30',
        },
        {
          generalLocation: 'Outset Island',
          detailedLocation: 'Savage Labyrinth - Floor 50',
        },
        {
          generalLocation: 'Dragon Roost Cavern',
          detailedLocation: 'First Room',
        },
        {
          generalLocation: 'Dragon Roost Cavern',
          detailedLocation: 'Alcove With Water Jugs',
        },
        {
          generalLocation: 'Dragon Roost Cavern',
          detailedLocation: "Bird's Nest",
        },
        {
          generalLocation: 'Outset Island',
          detailedLocation: 'Sunken Treasure',
        },
      ];
      Locations.initializeRaw(itemLocations, locationsList);

      state = new TrackerState();

      state.locationsChecked = {
        'Outset Island': {
          'Savage Labyrinth - Floor 30': false,
          'Savage Labyrinth - Floor 50': false,
          'Sunken Treasure': false,
        },
        'Dragon Roost Cavern': {
          'First Room': false,
          'Alcove With Water Jugs': true,
          "Bird's Nest": false,
        },
        'Forsaken Fortress': {
          'Phantom Ganon': false,
        },
        Mailbox: {
          'Letter from Aryll': false,
          'Letter from Tingle': false,
        },
      };
    });

    test('return a new state with the banned locations cleared for DRC', () => {
      const newState = state.clearBannedLocations('Dragon Roost Cavern', { includeAdditionalLocations: true });

      expect(newState.locationsChecked).toEqual({
        'Outset Island': {
          'Savage Labyrinth - Floor 30': false,
          'Savage Labyrinth - Floor 50': false,
          'Sunken Treasure': false,
        },
        'Dragon Roost Cavern': {
          'First Room': true,
          'Alcove With Water Jugs': true,
          "Bird's Nest": true,
        },
        'Forsaken Fortress': {
          'Phantom Ganon': false,
        },
        Mailbox: {
          'Letter from Aryll': false,
          'Letter from Tingle': false,
        },
      });
    });

    test('return a new state with the banned locations cleared for Outset', () => {
      const newState = state.clearBannedLocations('Outset Island', { includeAdditionalLocations: true });

      expect(newState.locationsChecked).toEqual({
        'Outset Island': {
          'Savage Labyrinth - Floor 30': true,
          'Savage Labyrinth - Floor 50': true,
          'Sunken Treasure': true,
        },
        'Dragon Roost Cavern': {
          'First Room': false,
          'Alcove With Water Jugs': true,
          "Bird's Nest": false,
        },
        'Forsaken Fortress': {
          'Phantom Ganon': false,
        },
        Mailbox: {
          'Letter from Aryll': false,
          'Letter from Tingle': false,
        },
      });
    });

    test('return a new state with the banned locations and mail cleared for FF when includeAdditionalLocations is true', () => {
      const newState = state.clearBannedLocations('Forsaken Fortress', { includeAdditionalLocations: true });

      expect(newState.locationsChecked).toEqual({
        'Outset Island': {
          'Savage Labyrinth - Floor 30': false,
          'Savage Labyrinth - Floor 50': false,
          'Sunken Treasure': false,
        },
        'Dragon Roost Cavern': {
          'First Room': false,
          'Alcove With Water Jugs': true,
          "Bird's Nest": false,
        },
        'Forsaken Fortress': {
          'Phantom Ganon': true,
        },
        Mailbox: {
          'Letter from Aryll': true,
          'Letter from Tingle': true,
        },
      });
    });

    test('return a new state with the banned locations but mail not cleared for FF when includeAdditionalLocations is false', () => {
      const newState = state.clearBannedLocations('Forsaken Fortress', { includeAdditionalLocations: false });

      expect(newState.locationsChecked).toEqual({
        'Outset Island': {
          'Savage Labyrinth - Floor 30': false,
          'Savage Labyrinth - Floor 50': false,
          'Sunken Treasure': false,
        },
        'Dragon Roost Cavern': {
          'First Room': false,
          'Alcove With Water Jugs': true,
          "Bird's Nest": false,
        },
        'Forsaken Fortress': {
          'Phantom Ganon': true,
        },
        Mailbox: {
          'Letter from Aryll': false,
          'Letter from Tingle': false,
        },
      });
    });
  });
});
