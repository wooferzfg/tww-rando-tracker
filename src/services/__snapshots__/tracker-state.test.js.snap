// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackerState default initializes the checked locations 1`] = `
{
  "Dragon Roost Cavern": {
    "Bird's Nest": false,
  },
  "Outset Island": {
    "Savage Labyrinth - Floor 30": false,
  },
}
`;

exports[`TrackerState default initializes the items 1`] = `
{
  "Bait Bag": 0,
  "Ballad of Gales": 1,
  "Beedle's Chart": 0,
  "Boat's Sail": 1,
  "Bombs": 0,
  "Boomerang": 0,
  "Cabana Deed": 0,
  "Chart for Angular Isles": 0,
  "Chart for Bird's Peak Rock": 0,
  "Chart for Boating Course": 0,
  "Chart for Bomb Island": 0,
  "Chart for Cliff Plateau Isles": 0,
  "Chart for Crescent Moon Island": 0,
  "Chart for Cyclops Reef": 0,
  "Chart for Diamond Steppe Island": 0,
  "Chart for Dragon Roost Island": 0,
  "Chart for Eastern Fairy Island": 0,
  "Chart for Eastern Triangle Island": 0,
  "Chart for Fire Mountain": 0,
  "Chart for Five-Eye Reef": 0,
  "Chart for Five-Star Isles": 0,
  "Chart for Flight Control Platform": 0,
  "Chart for Forest Haven": 0,
  "Chart for Forsaken Fortress": 0,
  "Chart for Four-Eye Reef": 0,
  "Chart for Gale Isle": 0,
  "Chart for Greatfish Isle": 0,
  "Chart for Headstone Island": 0,
  "Chart for Horseshoe Island": 0,
  "Chart for Ice Ring Isle": 0,
  "Chart for Islet of Steel": 0,
  "Chart for Mother and Child Isles": 0,
  "Chart for Needle Rock Isle": 0,
  "Chart for Northern Fairy Island": 0,
  "Chart for Northern Triangle Island": 0,
  "Chart for Outset Island": 0,
  "Chart for Overlook Island": 0,
  "Chart for Pawprint Isle": 0,
  "Chart for Private Oasis": 0,
  "Chart for Rock Spire Isle": 0,
  "Chart for Seven-Star Isles": 0,
  "Chart for Shark Island": 0,
  "Chart for Six-Eye Reef": 0,
  "Chart for Southern Fairy Island": 0,
  "Chart for Southern Triangle Island": 0,
  "Chart for Spectacle Island": 0,
  "Chart for Star Belt Archipelago": 0,
  "Chart for Star Island": 0,
  "Chart for Stone Watcher Island": 0,
  "Chart for Thorned Fairy Island": 0,
  "Chart for Three-Eye Reef": 0,
  "Chart for Tingle Island": 0,
  "Chart for Tower of the Gods": 0,
  "Chart for Two-Eye Reef": 0,
  "Chart for Western Fairy Island": 0,
  "Chart for Windfall Island": 0,
  "Command Melody": 0,
  "DRC Big Key": 0,
  "DRC Compass": 0,
  "DRC Dungeon Map": 0,
  "DRC Small Key": 0,
  "Deku Leaf": 0,
  "Delivery Bag": 0,
  "Din's Pearl": 0,
  "ET Big Key": 0,
  "ET Compass": 0,
  "ET Dungeon Map": 0,
  "ET Small Key": 0,
  "Earth God's Lyric": 0,
  "Empty Bottle": 0,
  "Entered Angular Isles Cave": 0,
  "Entered Bird's Peak Rock Cave": 0,
  "Entered Boating Course Cave": 0,
  "Entered Bomb Island Cave": 0,
  "Entered Cabana Labyrinth": 0,
  "Entered Cliff Plateau Isles Cave": 0,
  "Entered DRC": 0,
  "Entered Diamond Steppe Island Cave": 0,
  "Entered Dragon Roost Island Cave": 0,
  "Entered ET": 0,
  "Entered FF": 0,
  "Entered FW": 0,
  "Entered Fire Mountain Cave": 0,
  "Entered GT": 0,
  "Entered Horseshoe Island Cave": 0,
  "Entered Ice Ring Isle Cave": 0,
  "Entered Needle Rock Isle Cave": 0,
  "Entered Overlook Island Cave": 0,
  "Entered Pawprint Isle Chuchu Cave": 0,
  "Entered Pawprint Isle Wizzrobe Cave": 0,
  "Entered Rock Spire Isle Cave": 0,
  "Entered Savage Labyrinth": 0,
  "Entered Shark Island Cave": 0,
  "Entered Star Island Cave": 0,
  "Entered Stone Watcher Island Cave": 0,
  "Entered TotG": 0,
  "Entered WT": 0,
  "FF Compass": 0,
  "FF Dungeon Map": 0,
  "FW Big Key": 0,
  "FW Compass": 0,
  "FW Dungeon Map": 0,
  "FW Small Key": 0,
  "Farore's Pearl": 0,
  "Fill-Up Coupon": 0,
  "Ghost Ship Chart": 0,
  "Grappling Hook": 0,
  "Great Fairy Chart": 0,
  "Hero's Charm": 0,
  "Hookshot": 0,
  "Hurricane Spin": 0,
  "Iron Boots": 0,
  "Light Ring Chart": 0,
  "Maggie's Letter": 0,
  "Magic Armor": 0,
  "Moblin's Letter": 0,
  "Nayru's Pearl": 0,
  "Note to Mom": 0,
  "Octo Chart": 0,
  "Platform Chart": 0,
  "Power Bracelets": 0,
  "Progressive Bomb Bag": 0,
  "Progressive Bow": 0,
  "Progressive Magic Meter": 0,
  "Progressive Picto Box": 0,
  "Progressive Quiver": 0,
  "Progressive Shield": 1,
  "Progressive Sword": 1,
  "Progressive Wallet": 0,
  "Secret Cave Chart": 0,
  "Skull Hammer": 0,
  "Song of Passing": 1,
  "Spoils Bag": 0,
  "Submarine Chart": 0,
  "Telescope": 0,
  "Tingle Statue": 0,
  "Tingle Tuner": 0,
  "Tingle's Chart": 0,
  "TotG Big Key": 0,
  "TotG Compass": 0,
  "TotG Dungeon Map": 0,
  "TotG Small Key": 0,
  "Treasure Chart 1": 0,
  "Treasure Chart 10": 0,
  "Treasure Chart 11": 0,
  "Treasure Chart 12": 0,
  "Treasure Chart 13": 0,
  "Treasure Chart 14": 0,
  "Treasure Chart 15": 0,
  "Treasure Chart 16": 0,
  "Treasure Chart 17": 0,
  "Treasure Chart 18": 0,
  "Treasure Chart 19": 0,
  "Treasure Chart 2": 0,
  "Treasure Chart 20": 0,
  "Treasure Chart 21": 0,
  "Treasure Chart 22": 0,
  "Treasure Chart 23": 0,
  "Treasure Chart 24": 0,
  "Treasure Chart 25": 0,
  "Treasure Chart 26": 0,
  "Treasure Chart 27": 0,
  "Treasure Chart 28": 0,
  "Treasure Chart 29": 0,
  "Treasure Chart 3": 0,
  "Treasure Chart 30": 0,
  "Treasure Chart 31": 0,
  "Treasure Chart 32": 0,
  "Treasure Chart 33": 0,
  "Treasure Chart 34": 0,
  "Treasure Chart 35": 0,
  "Treasure Chart 36": 0,
  "Treasure Chart 37": 0,
  "Treasure Chart 38": 0,
  "Treasure Chart 39": 0,
  "Treasure Chart 4": 0,
  "Treasure Chart 40": 0,
  "Treasure Chart 41": 0,
  "Treasure Chart 5": 0,
  "Treasure Chart 6": 0,
  "Treasure Chart 7": 0,
  "Treasure Chart 8": 0,
  "Treasure Chart 9": 0,
  "Triforce Chart 1": 0,
  "Triforce Chart 2": 0,
  "Triforce Chart 3": 0,
  "Triforce Chart 4": 0,
  "Triforce Chart 5": 0,
  "Triforce Chart 6": 0,
  "Triforce Chart 7": 0,
  "Triforce Chart 8": 0,
  "Triforce Shard": 0,
  "WT Big Key": 0,
  "WT Compass": 0,
  "WT Dungeon Map": 0,
  "WT Small Key": 0,
  "Wind God's Aria": 0,
  "Wind Waker": 1,
  "Wind's Requiem": 1,
}
`;

exports[`TrackerState default initializes the items for locations 1`] = `
{
  "Dragon Roost Cavern": {
    "Bird's Nest": null,
  },
  "Outset Island": {
    "Savage Labyrinth - Floor 30": null,
  },
}
`;
